File: Day35/LowestCommonAncestor_BinTree.java
import java.util.*;

class TreeNode {
    int val;
    TreeNode left, right;
    TreeNode(int x) { val = x; }
}

public class LowestCommonAncestor_BinTree {
    public static TreeNode buildTree(String[] vals) {
        if (vals.length == 0 || vals[0].equals("null")) return null;
        TreeNode[] nodes = new TreeNode[vals.length];
        for (int i = 0; i < vals.length; i++) {
            if (!vals[i].equals("null")) nodes[i] = new TreeNode(Integer.parseInt(vals[i]));
        }
        int idx = 1;
        for (int i = 0; i < vals.length && idx < vals.length; i++) {
            if (nodes[i] != null) {
                if (idx < vals.length) nodes[i].left = nodes[idx++];
                if (idx < vals.length) nodes[i].right = nodes[idx++];
            }
        }
        return nodes[0];
    }
    public static TreeNode findNode(TreeNode root, int val) {
        if (root == null) return null;
        if (root.val == val) return root;
        TreeNode left = findNode(root.left, val);
        if (left != null) return left;
        return findNode(root.right, val);
    }
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        if (root == null || root == p || root == q) return root;
        TreeNode left = lowestCommonAncestor(root.left, p, q);
        TreeNode right = lowestCommonAncestor(root.right, p, q);
        if (left != null && right != null) return root;
        return left != null ? left : right;
    }
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        System.out.print("Enter tree nodes in level order (comma separated, use null for missing): ");
        String[] vals = sc.nextLine().split(",");
        TreeNode root = buildTree(vals);
        System.out.print("Enter value of p: ");
        int pVal = sc.nextInt();
        System.out.print("Enter value of q: ");
        int qVal = sc.nextInt();
        TreeNode p = findNode(root, pVal);
        TreeNode q = findNode(root, qVal);
        LowestCommonAncestor_BinTree sol = new LowestCommonAncestor_BinTree();
        TreeNode lca = sol.lowestCommonAncestor(root, p, q);
        System.out.println("LCA value: " + lca.val);
    }
}

File: Day35/LowestCommonAncestor_BinTree.py
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

def build_tree(values):
    if not values or values[0] is None:
        return None
    nodes = [TreeNode(v) if v is not None else None for v in values]
    kids = nodes[::-1]
    root = kids.pop()
    for node in nodes:
        if node:
            if kids: node.left = kids.pop()
            if kids: node.right = kids.pop()
    return root

def find_node(root, val):
    if not root:
        return None
    if root.val == val:
        return root
    left = find_node(root.left, val)
    if left:
        return left
    return find_node(root.right, val)

class Solution(object):
    def lowestCommonAncestor(self, root, p, q):
        if not root or root == p or root == q:
            return root
        left = self.lowestCommonAncestor(root.left, p, q)
        right = self.lowestCommonAncestor(root.right, p, q)
        if left and right:
            return root
        return left or right

if __name__ == "__main__":
    inp = input("Enter tree nodes in level order (comma separated, use null for missing): ")
    nodes = [int(x) if x != 'null' else None for x in inp.strip().split(',')]
    root = build_tree(nodes)
    p_val = int(input("Enter value of p: "))
    q_val = int(input("Enter value of q: "))
    p = find_node(root, p_val)
    q = find_node(root, q_val)
    lca = Solution().lowestCommonAncestor(root, p, q)
    print("LCA value:", lca.val)
